recipes/openai_textcat.py

```diff
@@ -0,0 +1,535 @@
+        to display in the Prodigy UI.
+        """
+        stream = prodigy.components.preprocess.add_tokens(nlp, stream, skip=True)  # type: ignore
+        for example in stream:
+            example = copy.deepcopy(example)
+            response = self._parse_response(example["openai"]["response"])
+            example["answer"] = response["answer"] == "accept"
+            example["meta"]["reason"] = response["reason"]
+            yield example
+
+    def filter_suggestions(
+        self, stream: Iterable[Dict], negative_bias: float
+    ) -> Iterable[Dict]:
+        """Filter the examples based on some negative_bias"""
+        for example in stream:
+            if example.get("answer") != "accept":
+                rng = random.uniform(0, 1)
+                if rng <= negative_bias:
+                    eg = example
+                else:
+                    continue
+            else:
+                eg = example
+            yield prodigy.util.set_hashes(eg)
+
+    def _get_textcat_prompt(
+        self, text: str, label: str, examples: List[PromptExample]
+    ) -> str:
+        """Generate a prompt for binary text categorization."""
+        return self.prompt_template.render(text=text, label=label)
+
+    def _get_textcat_response(self, prompts: List[str], delay: int = 0) -> List[str]:
+        headers = {
+            "Authorization": f"Bearer {self.openai_api_key}",
+            "OpenAI-Organization": self.openai_api_org,
+            "Content-Type": "application/json",
+        }
+        time.sleep(delay)
+        r = _retry429(
+            lambda: client.post(
+                "https://api.openai.com/v1/completions",
+                headers=headers,
+                json={
+                    "model": self.model,
+                    "prompt": prompts,
+                    "temperature": self.openai_temperature,
+                    "max_tokens": self.openai_max_tokens,
+                },
+            ),
+            n=self.openai_n,
+            timeout_s=self.openai_timeout_s,
+        )
+        r.raise_for_status()
+        responses = r.json()
+        return [responses["choices"][i]["text"] for i in range(len(prompts))]
+
+    def _parse_response(self, text: str) -> Dict[str, str]:
+        """Interpret OpenAI's TextCat response. It's supposed to be
+        a list of lines, with each line having the form:
+
+        answer: <accept/reject for binary or the label for multilabel>
+        reason: <string>
+
+        However, there's no guarantee that the model will give
+        us well-formed output. It could say anything, it's an LM.
+        So we need to be robust.
+        """
+        output = {}
+        for line in text.strip().split("\n"):
+            if line and ":" in line:
+                key, value = line.split(":", 1)
+                output[key.strip().lower()] = value.strip()
+        return output
+
+
+@prodigy.recipe(
+    # fmt: off
+    "textcat.openai.suggest",
```

Comment:
I tried running this recipe and wondered if I would use the `fetch` command instead. 

I imagine this is somewhat a personal perference, but it seems easier to just download everything upfront first with the `fetch` recipe and to then just filter and only look at the examples where OpenAI had the class that I was interested in. This would work for both binary scenarios, as well as multi-label ones. 

@ljvmiranda921 what's your experience with this?